### ISSUE TITLE
[0.7-M4] Inserting text via user keyboard input inserts text in reverse order because caret position doesn't advance forward after input

### ISSUE THREAD

**Issue by DaveJarvis** [TomasMikula/RichTextFX#472]
**Replicate**
1. Run the Java Keywords demo
1. Start typing

**Expected Results**
The caret advances with each character typed.

**Actual Results**
The caret retains its position, which causes the typed text to be entered backwards.

![java-keywords-demo](https://cloud.githubusercontent.com/assets/2131950/24387596/d8c890d4-132a-11e7-8c7d-22645f98c2fd.png)

**Comment by JordanMartinez** [comment]
I'm guessing you're using the new 0.7-M4 release? That's probably a bug from #468. :-/

**Comment by JordanMartinez** [comment]
Nope. I was wrong. The bug start showing up in #463. After I moved the model-API to ESD, every time one types a letter, it moves the caret to the right twice so that the caret is one character farther than it should be. After I moved the view API to the area, the behavior reverted to what @DaveJarvis just now described in this issue.

**Comment by DaveJarvis** [comment]
Correct, this is in the 0.7-M4 release. 0.7-M3 worked as expected.

**Comment by JordanMartinez** [comment]
I've updated the release notes to note this issue.

**Comment by JordanMartinez** [comment]
@TomasMikula Would it be possible to re-release the 0.7-M4 release using the updated master branch that accounts for this issue? Or would this require another release?

**Comment by JordanMartinez** [comment]
Just fixed it. Dave, could you try it out to insure there are no other
regressions?

Although I already tagged you in the issue, is it possible, Tomas, to
re-release the 0.7-M4 release with the current master branch? Or does this
require a new release?

Blessings,
Jordan

**Comment by DaveJarvis** [comment]
@JordanMartinez, once the new release is available, I'll try it out.

I'd suggest making a new release, 0.7-M5. (Mostly to avoid caching issues.)

**Comment by TomasMikula** [comment]
I pushed a new SNAPSHOT release, can you try that out?

I think it used to be possible to override a release, but yeah, there can be caching issues.

**Comment by JordanMartinez** [comment]
:-/ Dang...

**Comment by DaveJarvis** [comment]
SNAPSHOT release has resolved this issue.

**Comment by JordanMartinez** [comment]
Can a new milestone release be made then?

@TomasMikula Besides the bugs and dependency bugs tagged in this project's list of issues, is there anything else that needs to be done before a stable `0.7` release can be made?

**Comment by JordanMartinez** [comment]
It is possible to fix this regression in the `0.7-M4` release by overriding the `replace` method like so:
````java
public void replace(int from, int to, StyledDocument<PS, SEG, S> document) {
    super.replace(from, to, document);
    int newCaretPos = getCaretPosition() + document.length();
    selectRange(newCaretPos, newCaretPos);
}
````

**Comment by TomasMikula** [comment]
Released 0.7-M5. As for 0.7, I don't know, mainly be it more stable, API-wise and bug-wise.

**Comment by JordanMartinez** [comment]
Ok. Bug-wise, there's 2 RTFX-specific bugs that can be fixed somewhat quickly (#423 & #362). I'm not sure what is causing the problem with #265 and I can't test (and thus fix) it because I don't use/have a Mac. Then there's the Flowless-related bugs (#412, #390, and probably #260).

How could the API be made more stable?
